### ISSUE TITLE
Translator should trigger downstream mining job clean-up for Extended Mining Jobs

### ISSUE THREAD

**Issue by adoerr** [stratum-mining/stratum#797]
Currently [handle_new_exteded_mining_job_](https://github.com/stratum-mining/stratum/blob/361f12ba72053422d64bfcb7ab010d5d45e1f0a5/roles/translator/src/lib/proxy/bridge.rs#L402-L452) is setting the `clean_job` parameter of [clean_job](https://github.com/stratum-mining/stratum/blob/361f12ba72053422d64bfcb7ab010d5d45e1f0a5/roles/translator/src/lib/proxy/next_mining_notify.rs#L15-L58) to `false` always.

This is to rigid and should only be done, if the downstream client does signal support for Extended Mining Jobs. Perhaps a default value of `true` is more sensible.

For additional context please find see the [Discord message](https://discord.com/channels/950687892169195530/958814900770205739/1219201652322730094)  message below:

--- 

First of all, please take everything I write with a grain of salt. I'm a complete stratum noob :wink:

Basically you are right, the "delay" refers to the fact that upstream is sending a `mining.notify` for job-id 7 while downstream is sending a `mining.submit` for job-id 4 *after* the notification for job-id 7 has been sent out.This will result in upstream reject the job-id 4 share submission. #791 is about handling this error condition properly.

The interesting question is, why both, `CPUminer` and `bitaxe` are submitting stale shares? SV1 `mining.notify` has a `clean_jobs` parameter which should prevent the above scenario or at least result in only rather infrequent invalid share submissions. Looking at the logs provided with [#788](https://github.com/stratum-mining/stratum/issues/788) it looks like the invalid share submission happens in conjunction with so called Extended Mining Jobs.

Looking at the translator code reveals that [create_notify()](https://github.com/stratum-mining/stratum/blob/361f12ba72053422d64bfcb7ab010d5d45e1f0a5/roles/translator/src/lib/proxy/next_mining_notify.rs#L15-L58) is responsible to create SV1 `mining.notify` requests. In the signature of create_notify() we will find the `clean_jobs` parameter talked about earlier.

[handle_new_prev_hash_()](https://github.com/stratum-mining/stratum/blob/361f12ba72053422d64bfcb7ab010d5d45e1f0a5/roles/translator/src/lib/proxy/bridge.rs#L304-L361) is calling create_notify() with `clean_jobs` set to `true`. So this should trigger downstream to clean up stale job-ids, which at least the `bitaxe` FW is doing.

However, [handle_new_extended_mining_job_()](https://github.com/stratum-mining/stratum/blob/361f12ba72053422d64bfcb7ab010d5d45e1f0a5/roles/translator/src/lib/proxy/bridge.rs#L402-L452) is *always* setting `clean_jobs` to `false` and I think. this is wrong. If a downstream client in a group channel doesn't signal support for extended mining jobs, `clean_jobs` should be `true` IMO.

---

**Comment by adoerr** [comment]
IMO, if we do not set `clean_jobs` to true we will basically kind of DDOS Sv1 downstream clients. A substantial part of the hashing capacity will be used for invalid job-ids and stale share submission :shrug:

**Comment by GitGab19** [comment]
> IMO, if we do not set `clean_jobs` to true we will basically kind of DDOS Sv1 downstream clients. A substantial part of the hashing capacity will be used for invalid job-ids and stale share submission 🤷

Could you check the behaviour of some pools, with your bitaxe? 
I just tried with cpu-miner pointing it to some pools, and all of them sent me `mining.notify` with `clean_jobs=false`, setting it to `true` only when a new block is found.

**Comment by adoerr** [comment]
Yep, I ran some tests with two ultras and at least 3/4th of the share submissions were stale. Setting `clean_jobs` to `true` results in *very* infrequent stale share submission. Like once in a couple of minutes.

**Comment by GitGab19** [comment]
> Yep, I ran some tests with two ultras and at least 3/4th of the share submissions were stale. Setting `clean_jobs` to `true` results in _very_ infrequent stale share submission. Like once in a couple of minutes.

Yeah that's for sure, but I meant trying with some real pools out there. To see how they behave and how they manage that parameter. Becuase I just tried and they all did the same way we are managing it rn

**Comment by adoerr** [comment]
Ah sorry :facepalm: So far I did only ran the ultras against public-pool.io, which is Sv1 only.

**Comment by GitGab19** [comment]
> Ah sorry 🤦 So far I did only ran the ultras against public-pool.io, which is Sv1 only.

I just pointed a cpu-miner to it and it seems that also public-pool.io sets it to false if it's not related to a new block

**Comment by adoerr** [comment]
Maybe I misunderstand something. My reasoning is based on [5.3.16 NewExtendedMiningJob (Server -> Client)](https://stratumprotocol.org/specification/05-Mining-Protocol/#5316-newextendedminingjob-server---client). Especially

`
 A proxy MUST translate the message for all downstream channels belonging to the group which don’t signal that they accept extended mining jobs in the SetupConnection message (intended and expected behavior for end mining devices).
`

Since a Sv1 downstream client implicitly never signals support  for accepting extended mining jobs my understanding is that `handle_new_extended_mining_job_()` and `handle_new_prev_hash_()` basically have to behave the same way. with regards to `clean_jobs`.

**Comment by GitGab19** [comment]
Closing this as stale.
